balancer/rls/picker.go

```diff
@@ -164,29 +175,46 @@ func (p *rlsPicker) delegateToChildPoliciesLocked(dcEntry *cacheEntry, info bala
 			// X-Google-RLS-Data header.
 			res, err := state.Picker.Pick(info)
 			if err != nil {
-				return res, err
+				return res, func() {
+					targetPicksMetric.Record(p.lb.bopts.MetricsRecorder, 1, p.lb.bopts.Target.String(), p.rlsServerTarget, cpw.target, "fail")
+				}, err
 			}
+
 			if res.Metadata == nil {
 				res.Metadata = metadata.Pairs(rlsDataHeaderName, dcEntry.headerData)
 			} else {
 				res.Metadata.Append(rlsDataHeaderName, dcEntry.headerData)
 			}
-			return res, nil
+			return res, func() {
+				targetPicksMetric.Record(p.lb.bopts.MetricsRecorder, 1, p.lb.bopts.Target.String(), p.rlsServerTarget, cpw.target, "complete")
+			}, nil
 		}
 	}
+
 	// In the unlikely event that we have a cache entry with no targets, we end up
 	// queueing the RPC.
-	return balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	return balancer.PickResult{}, func() {}, balancer.ErrNoSubConnAvailable
 }
 
 // useDefaultPickIfPossible is a helper method which delegates to the default
-// target if one is configured, or fails the pick with the given error.
-func (p *rlsPicker) useDefaultPickIfPossible(info balancer.PickInfo, errOnNoDefault error) (balancer.PickResult, error) {
+// target if one is configured, or fails the pick with the given error. Returns
+// a function to be invoked to record metrics.
+func (p *rlsPicker) useDefaultPickIfPossible(info balancer.PickInfo, errOnNoDefault error) (balancer.PickResult, func(), error) {
 	if p.defaultPolicy != nil {
 		state := (*balancer.State)(atomic.LoadPointer(&p.defaultPolicy.state))
-		return state.Picker.Pick(info)
+		res, err := state.Picker.Pick(info)
+		pr := "complete"
+		if err != nil {
+			pr = "fail"
+		}
```

Comment:
I don't think this is as simple. In C-core, the picker returns the chosen subchannel and a pick-result. The picker can return one of the following result types:

1. Complete: The pick is complete, and the picker returns a subchannel to send the RPC on.  It may also optionally return some headers to add to the RPC or request notification when the RPC ends.

2. Queue: The picker does not currently have a subchannel available to send the RPC on, but it is working on getting one (e.g., if there is a connection attempt currently in flight).  The channel will queue the RPC, and it will retry the pick when the LB policy returns a new picker.

3. Fail: The picker has failed to connect to all possible subchannels and has nowhere to send the RPC.  In the channel, if the RPC is wait_for_ready, it will be queued; otherwise, the RPC attempt will fail (although there may be additional attempts made by the retry code).

4. Drop: This is similar to "Fail", except that (a) even wait_for_ready RPCs will be failed and (b) no more retry attempts will be made.

In our metrics, we are not emitting anything for `Queue`, but we are for the other three. If you see the documentation for the `Pick` method in the `balancer` package, you will see:
```
	// If an error is returned:
	//
	// - If the error is ErrNoSubConnAvailable, gRPC will block until a new
	//   Picker is provided by the balancer (using ClientConn.UpdateState).
	//
	// - If the error is a status error (implemented by the grpc/status
	//   package), gRPC will terminate the RPC with the code and message
	//   provided.
	//
	// - For all other errors, wait for ready RPCs will wait, but non-wait for
	//   ready RPCs will be terminated with this error's Error() string and
	//   status code Unavailable.
```

This means that for us, the pick result can be deduced as follows:
- `err == nil`: pick_result is "complete"
- `err == ErrNoSubConnAvailable`: pick_result is "queue"
- `err` is not a `status` error: pick_result is "fail"
  - although, i believe we will not be able to differentiate here between `wait_for_ready` RPCs and the others. This needs to be clarified with someone to see if this is OK.
  - @dfawley : What do you think? Do you think it would be an OK idea to add a `FailFast` field to `PickInfo`? The `picker_wrapper` has this information, it just doesn't pass it to the `Pick` method.
- `err` is a `status` error: pick_result is "drop"